### PR TITLE
Support HPA

### PR DESCRIPTION
**Pull request by loganhz**
Proposed changes
======
Support HPA

Types of changes
======
- New feature (non-breaking change which adds functionality)

Linked Issues
======
https://github.com/rancher/rancher/issues/16214
 

Further comments
======

### CODE REVIEW

File: app/components/cru-hpa/template.hbs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+          namespace=namespace
+          errors=namespaceErrors
+          registerHook=(action "registerHook")
+          editing=editing
+        }}
+      </div>
+    </div>
+    <div class="row">
+      <div class="col span-6">
+        <label class="pb-5 acc-label desc-label">
```
Comment (westlywright) [inline app/components/cru-hpa/template.hbs]: Add for attribute and matching id to select/input

File: app/components/cru-hpa/template.hbs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+          value=model.workloadId
+          showProTip=false
+          editing=editing
+          disabled=isView
+        }}
+      </div>
+    </div>
+    <div class="row">
+      <div class="col span-6">
+        <label class="pb-5 acc-label desc-label">
```
Comment (westlywright) [inline app/components/cru-hpa/template.hbs]: Add for attribute and matching id to select/input

File: app/components/cru-hpa/template.hbs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        {{input-integer
+          min=0
+          max=model.maxReplicas
+          value=model.minReplicas
+          classNames="form-control"
+          placeholder=(t "cruHpa.minReplicas.placeholder")
+        }}
+      </div>
+      <div class="col span-6">
+        <label class="pb-5 acc-label desc-label">
```
Comment (westlywright) [inline app/components/cru-hpa/template.hbs]: Add for attribute and matching id to select/input

File: app/components/hpa-metric-row/component.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+  editing: true,
+  metric:  null,
+
+  typeChoices:     METRICS_TYPES,
+  resourceChoices: RESOURCE_TYPES,
+
+  actions: {
+    removeMetric(metric) {
+      this.sendAction('removeMetric', metric);
```
Comment (westlywright) [inline app/components/hpa-metric-row/component.js]: `sendAction` is being deprecated, use closure actions instead

File: app/components/hpa-metric-row/template.hbs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+          {{#if metric.current.value}}
+            {{t "cruHpa.metrics.target.options.value"}}: {{metric.current.value}}
+          {{/if}}
+        </div>
+      </div>
+    </div>
+  {{/unless}}
+  <div class="row">
+    <div class="col span-6">
+      <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: app/components/hpa-metric-row/template.hbs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+      {{new-select
+        classNames="form-control"
+        content=typeChoices
+        value=metric.type
+        localizedLabel=true
+        disabled=(not editing)
+      }}
+    </div>
+    <div class="col span-6">
+      <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: app/components/hpa-metric-row/template.hbs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+            value=metric.name
+            placeholder=(t "cruHpa.metrics.name.placeholder")
+          }}
+        {{/input-or-display}}
+      {{/if}}
+    </div>
+  </div>
+  <div class="row">
+    <div class="col span-6">
+      <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: app/components/hpa-metric-row/template.hbs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+      {{new-select
+        classNames="form-control"
+        content=targetChoices
+        value=metric.target.type
+        localizedLabel=true
+        disabled=(not editing)
+      }}
+    </div>
+    <div class="col span-6">
+      <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: app/components/hpa-metric-row/template.hbs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            placeholder=(t (if (eq metric.name "cpu") "cruHpa.metrics.value.cpuPlaceholder" "cruHpa.metrics.value.placeholder"))
+          }}
+        {{/input-or-display}}
+      {{/if}}
+    </div>
+  </div>
+  {{#if (eq metric.type "Object")}}
+    <div class="row">
+      <div class="col span-6">
+        <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: app/components/hpa-metric-row/template.hbs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+          {{input
+            class="form-control"
+            type="text"
+            value=metric.describedObject.apiVersion
+            placeholder=(t "cruHpa.metrics.describedObject.apiVersion.placeholder")
+          }}
+        {{/input-or-display}}
+      </div>
+      <div class="col span-6">
+        <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: app/components/hpa-metric-row/template.hbs [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+            type="text"
+            value=metric.describedObject.kind
+            placeholder=(t "cruHpa.metrics.describedObject.kind.placeholder")
+          }}
+        {{/input-or-display}}
+      </div>
+    </div>
+    <div class="row">
+      <div class="col span-6">
+        <label class="pb-5 acc-label">
```
Comment (westlywright) [inline app/components/hpa-metric-row/template.hbs]: Add `for` attribute and matching `id` to select/input

File: lib/shared/addon/components/form-match-expressions/component.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        key:      '',
+        operator: EXISTS,
+      });
+
+      next(() => {
+        if ( this.isDestroyed || this.isDestroying ) {
+          return;
+        }
+
+        let elem = this.$('INPUT.key').last()[0];
```
Comment (westlywright) [inline lib/shared/addon/components/form-match-expressions/component.js]: `this.$` will be deprecated soon, import `$` from jquery and use it directly.

File: app/authenticated/project/hpa/detail/edit/route.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import Route from '@ember/routing/route';
+import { get } from '@ember/object';
+import { hash } from 'rsvp';
+
+export default Route.extend({
+  model() {
+    const store = get(this, 'store');
+    const original = this.modelFor('authenticated.project.hpa.detail').hpa;
```
[the remark this answers is not in the document]
Comment (loganhz) [reply app/authenticated/project/hpa/detail/edit/route.js]: Tested.
If hpa doesn't exist, `hpa/detail/route` will handle it.

Users will see the general rancher 404 page.

```
NotFound (404)
horizontalpodautoscalers.autoscaling "hpa-x4sp3" not found
```

**Comment by loganhz** [comment]
Updated. Thanks!

**Comment by cjellick** [comment]
backend is in

**Pull request merged**